ssl/quic/quic_rx_depack.c

```diff
@@ -0,0 +1,932 @@
+                                        OSSL_ACKM_RX_PKT *ackm_data)
+{
+    OSSL_QUIC_FRAME_STOP_SENDING frame_data;
+    QUIC_STREAM *stream = NULL;
+    int stream_type = 0;
+
+    if (!ossl_quic_wire_decode_frame_stop_sending(pkt, &frame_data))
+        return 0;
+
+    if (mode == collect_ack_data) {
+        /* This frame makes the packet ACK eliciting */
+        ackm_data->is_ack_eliciting = 1;
+        return 1;
+    }
+
+    if ((stream = ssl_get_stream(connection, frame_data.stream_id)) == NULL)
+        return 0;
+    stream_type = ssl_get_stream_type(stream);
+
+    ssl_close_stream(stream, SHUT_RD); /* Reuse shutdown(2) symbols */
+    if ((stream_type & SSL_STREAM_TYPE_R) != 0) {
+        QUICfatal(connection, QUIC_STREAM_STATE_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    return 1;
+}
+
+static int depack_do_frame_crypto(PACKET *pkt, QUIC_CONNECTION *connection,
+                                  enum depack_modes mode,
+                                  OSSL_ACKM_RX_PKT *ackm_data)
+{
+    OSSL_QUIC_FRAME_CRYPTO frame_data;
+
+    if (!ossl_quic_wire_decode_frame_crypto(pkt, &frame_data))
+        return 0;
+
+    if (mode == collect_ack_data) {
+        /* This frame makes the packet ACK eliciting */
+        ackm_data->is_ack_eliciting = 1;
+        return 1;
+    }
+
+    /* ADD CODE to send |frame_data.data| to the handshake manager */
+
+    return 1;
+}
+
+static int depack_do_frame_new_token(PACKET *pkt, QUIC_CONNECTION *connection,
+                                     enum depack_modes mode,
+                                     OSSL_ACKM_RX_PKT *ackm_data)
+{
+    const uint8_t *token;
+    size_t token_len;
+
+    if (!ossl_quic_wire_decode_frame_new_token(pkt, &token, &token_len))
+        return 0;
+
+    if (mode == collect_ack_data) {
+        /* This frame makes the packet ACK eliciting */
+        ackm_data->is_ack_eliciting = 1;
+        return 1;
+    }
+
+    /* ADD CODE to send |token| to the session manager */
+
+    return 1;
+}
+
+static int depack_do_frame_stream(PACKET *pkt, QUIC_CONNECTION *connection,
+                                  OSSL_QRX_PKT_WRAP *parent_pkt,
+                                  enum depack_modes mode,
+                                  OSSL_ACKM_RX_PKT *ackm_data)
+{
+    OSSL_QUIC_FRAME_STREAM frame_data;
+    QUIC_STREAM *stream;
+
+    if (!ossl_quic_wire_decode_frame_stream(pkt, &frame_data))
+        return 0;
+
+    if (mode == collect_ack_data) {
+        /* This frame makes the packet ACK eliciting */
+        ackm_data->is_ack_eliciting = 1;
+        return 1;
+    }
+
+    /*
+     * ASSUMPTION: ssl_get_stream() gets a QUIC_STREAM from a
+     * QUIC_CONNECTION by stream ID.
+     */
+    if ((stream = ssl_get_stream(connection, frame_data.stream_id)) == NULL)
+        return 0;
+    if (!ssl_queue_data(stream, parent_pkt, frame_data.data, frame_data.len))
+        return 0;
+
+    /* ADD CODE to deal with |frame_data.is_fin| */
```

Comment:
I suppose the easiest thing would be to leave this up to the stream to handle that flag. I.e., just pass it to the ssl_queue_data() as an argument.

Reply:
It's possible.  I'm not ready to assume that's the way things will go, though.

Reply:
> > It's possible. I'm not ready to assume that's the way things will go, though.
> 
> Why not? I assume the stream receive buffer would be the only module to know how to handle it, including the possible protocol violations.

I imagine that it could as well be the depacketizer's responsibility to call something like ssl_stream_shutdown() with appropriate arguments if the FIN bit is set.  It depends on how the ssl API ends up being written.

Reply:
Well, that call would go to the rcv stream buffers anyway. But it would have to infer the last buffered stream frame somehow. Which might not be the actual last frame as per the offset+length received if the peer violates the protocol. The fin bit must be always associated with the stream frame data to properly detect the violation.

Reply:
Actually I see another problem with this prototype - the offset is of course needed by the stream too.

Reply:
Ok, I'm passing both.